File: go516/m516.go
package main

type CPU struct {
	reg       *Registers
	ram       *Memory
	instCount int
}

const (
	ACU, _ = iota, iota
	ADD, ONE
	SUB, NIL
	AND, ALL
	OR, RSH
	Aptr, _
	A, _
	PlusA, Aminus
	M0, _
	M1, _
	M2, _
	M3, _
	M4, _
	M5, _
	M6, _
	M7, _
	M8, _
	M9, _
	M10, _
	M11, _
	M12, _
	M13, _
	M14, _
	M15, _
	M16, _
	M17, _
	M18, _
	M19, _
	M20, _
	M21, _
	M22, _
	M23, _
	M24, _
	M25, _
	M26, _
	M27, _
	M28, _
	M29, _
	M30, _
	M31, _
	XOR, INV
	MinusA, Aplus
	HIGH, _
	LOW, _
	Bptr, _
	B, _
	PlusB, Bminus
	PC, _
	PCZ, LIT
	PCC, CONF
	RET, _
	Mptr, _
)

const CARRY = 0x80

func (c *CPU) Decode(instruction Word) {
	format := instruction >> 14

	if format == 0 {
		c.copy(Addr((instruction>>7)&0x7f), Addr(instruction&0x7f))
	} else if format == 1 {
		c.literal(Addr((instruction>>12)&0x03), instruction&0xfff)
	} else {
		c.copy(Addr((instruction>>12)&0x07), Addr((instruction>>8)&0x0f))
		c.copy(Addr((instruction>>4)&0x0f), Addr(instruction&0x0f))
	}
}

func (c *CPU) literal(dst Addr, val Word) {
	c.instCount++
	c.reg.write[dst](val)
}

func (c *CPU) copy(dst Addr, src Addr) {
	c.instCount++
	c.reg.write[dst](c.reg.read[src]())
}

func (c *CPU) PC() Word {
	return c.reg.value[PC]
}

func (c *CPU) GOTO(address Word) {
	c.reg.value[PC] = address
}

func (c *CPU) Exec() {
	pc := c.PC()
	instruction := c.ram.get(pc)
	c.GOTO(pc + 1)
	c.Decode(instruction)
}

func NewCPU() *CPU {
	var reg = NewRegisters()
	var RAM = NewMemory()

	reg.value[CARRY] = 0
	reg.Plain(ACU)
	reg.R(ONE, func() Word { return 1 })
	reg.R(NIL, func() Word { return 0 })
	reg.R(ALL, func() Word { return 0xffff })
	reg.W(ADD, func(v Word) {
		var calc int32
		calc = int32(reg.value[ACU]) + int32(v)
		if calc > 0xffff {
			reg.value[CARRY] = 1
		} else {
			reg.value[CARRY] = 0
		}
		reg.value[ACU] = Word(calc)
	})
	reg.W(SUB, func(v Word) {
		var calc int32
		calc = int32(reg.value[ACU]) - int32(v)
		if calc < 0 {
			reg.value[CARRY] = 0
		} else {
			reg.value[CARRY] = 1
		}
		reg.value[ACU] = Word(calc)
	})
	reg.W(AND, func(v Word) { reg.value[ACU] &= v })
	reg.R(RSH, func() Word { return reg.value[ACU] >> 1 })
	reg.W(OR, func(v Word) { reg.value[ACU] |= v })
	reg.Plain(Aptr)
	reg.R(INV, func() Word { return reg.value[ACU] ^ 0xffff })
	reg.W(XOR, func(v Word) { reg.value[ACU] ^= v })
	reg.Plain(Bptr)
	reg.Plain(PC)
	reg.W(PC, func(v Word) {
		reg.value[RET] = reg.value[PC]
		reg.value[PC] = v
	})
	reg.Plain(RET)
	reg.Plain(Mptr)
	reg.W(Mptr, func(v Word) { reg.value[Mptr] = v & 0xffe0 })
	reg.R(A, func() Word { return RAM.get(reg.value[Aptr]) })
	reg.W(A, func(v Word) { RAM.set(reg.value[Aptr], v) })
	reg.R(Aminus, func() Word {
		reg.value[Aptr]--
		return RAM.get(reg.value[Aptr] + 1)
	})
	reg.W(PlusA, func(v Word) {
		reg.value[Aptr]++
		RAM.set(reg.value[Aptr], v)
	})
	reg.R(Aplus, func() Word {
		reg.value[Aptr]++
		return RAM.get(reg.value[Aptr] - 1)
	})
	reg.W(MinusA, func(v Word) {
		reg.value[Aptr]--
		RAM.set(reg.value[Aptr], v)
	})
	reg.R(HIGH, func() Word {
		return RAM.get(reg.value[Aptr]) >> 8
	})
	reg.W(HIGH, func(v Word) {
		mvalue := RAM.get(reg.value[Aptr]) & 0x00ff
		RAM.set(reg.value[Aptr], mvalue|(v<<8))
	})
	reg.R(LOW, func() Word {
		return RAM.get(reg.value[Aptr]) & 0xff
	})
	reg.W(LOW, func(v Word) {
		mvalue := RAM.get(reg.value[Aptr]) & 0xff00
		RAM.set(reg.value[Aptr], mvalue|(v&0x00ff))
	})
	reg.R(B, func() Word { return RAM.get(reg.value[Bptr]) })
	reg.W(B, func(v Word) { RAM.set(reg.value[Bptr], v) })
	reg.R(Bminus, func() Word {
		reg.value[Bptr]--
		return RAM.get(reg.value[Bptr] + 1)
	})
	reg.W(PlusB, func(v Word) {
		reg.value[Bptr]++
		RAM.set(reg.value[Bptr], v)
	})
	reg.W(PCZ, func(v Word) {
		if reg.value[ACU] == 0 {
			reg.write[PC](v)
		}
	})
	reg.R(LIT, func() Word {
		reg.value[PC]++
		return RAM.get(reg.value[PC] - 1)
	})
	reg.W(PCC, func(v Word) {
		if reg.value[CARRY] != 0 {
			reg.write[PC](v)
		}
	})
	reg.R(CONF, func() Word { return 0x0001 })

	for i, m := 0x08, 0; i <= 0x27; i, m = i+1, m+1 {
		reg.R(Addr(i), func(offset Word) func() Word {
			return func() Word {
				return RAM.get(reg.value[Mptr] + offset)
			}
		}(Word(m)))

		reg.W(Addr(i), func(offset Word) func(Word) {
			return func(v Word) {
				RAM.set(reg.value[Mptr]+offset, v)
			}
		}(Word(m)))
	}

	return &CPU{
		reg: reg,
		ram: RAM,
	}
}

File: m516/c516/m516.c
#include "registers.h"
#include "memory.h"
#include "datatypes.h"
#include <stdio.h>

#define ACU    0x00
#define ADD    0x01
#define ONE    0x01
#define SUB    0x02
#define NIL    0x02
#define AND    0x03
#define ALL    0x03
#define RSH    0x04
#define OR     0x04
#define Aptr   0x05
#define A      0x06
#define PlusA  0x07
#define Aminus 0x07
#define M(x)   0x08 + x
#define XOR    0x28
#define INV    0x28
#define MinusA 0x29
#define Aplus  0x29
#define HIGH   0x2a
#define LOW    0x2b
#define Bptr   0x2c
#define B      0x2d
#define PlusB  0x2e
#define Bminus 0x2e
#define PC     0x2f
#define PCZ    0x30
#define LIT    0x30
#define PCC    0x31
#define CONF   0x31
#define RET    0x32
#define Mptr   0x33
#define BANK   0x34
#define BNKA   0x35

int instCount;

int carry () {
	int32_t sum;
	
	sum = Val[A] + Val[B];
	
	if (sum > 0xffff) {
		return 1;
	}
	return 0;
}

void writeA (word val) {
	Val[A] = val;
}
word readA () {
	return Val[A];
}

void writeB (word val) {
	Val[B] = val;
}
word readB () {
	return Val[B];
}

word add () {
	return Val[A] + Val[B];
}
word and () {
	return Val[A] & Val[B];
}
word or () {
	return Val[A] | Val[B];
}
word xor () {
	return Val[A] ^ Val[B];
}
word rsh () {
	return Val[A] >> 1;
}
word lit () {
	word ret;
	ret = RAM[Val[PC]];
	Val[PC]++;
	return ret;
}

void pc (word val) {
	Val[PC] = val;
}
void pcz (word val) {
	if (!(Val[A]+Val[B])) {
		Val[PC] = val;
	}
}
void pcn (word val) {
	if (Val[A]+Val[B]) {
		Val[PC] = val;
	}
}
void pcc (word val) {
	if (carry()) {
		Val[PC] = val;
	}
}
void pczc (word val) {
	if (!(Val[C])) {
		Val[PC] = val;
	}
}
void pczd (word val) {
	if (!(Val[D])) {
		Val[PC] = val;
	}
}

word readCP () {
	return Val[CP];
}
void writeCP (word val) {
	Val[CP] = val;
}
word readC () {
	return RAM[Val[CP]];
}
void writeC (word val) {
	setRAM(Val[CP], val);
}
word c_dec () {
	word ret;
	ret = RAM[Val[CP]];
	Val[CP]--;
	return ret;
}
void inc_c (word val) {
	Val[CP]++;
	setRAM(Val[CP], val);
}
word c_inc () {
	word ret;
	ret = RAM[Val[CP]];
	Val[CP]++;
	return ret;
}
void dec_c (word val) {
	Val[CP]--;
	setRAM(Val[CP], val);
}

word readDP () {
	return Val[DP];
}
void writeDP (word val) {
	Val[DP] = val;
}
word readD () {
	return RAM[Val[DP]];
}
void writeD (word val) {
	setRAM(Val[DP], val);
}
word dec_d () {
	Val[DP]--;
	return RAM[Val[DP]];
}
void d_inc (word val) {
	setRAM(Val[DP], val);
	Val[DP]++;
}
word inc_d () {
	Val[DP]++;
	return RAM[Val[DP]];
}
void d_dec (word val) {
	setRAM(Val[DP], val);
	Val[DP]--;
}

void initCPU () {
	initRAM();
	initRegisters();
	instCount = 0;
	
	R[A] = readA;
	W[A] = writeA;
	R[B] = readB;
	W[B] = writeB;
	R[ADD] = add;
	W[PC] = pc;
	R[AND] = and;
	W[PCZ] = pcz;
	R[OR] = or;
	W[PCN] = pcn;
	R[XOR] = xor;
	W[PCC] = pcc;
	R[RSH] = rsh;
	W[PCZC] = pczc;
	R[LIT] = lit;
	W[PCZD] = pczd;
	R[CP] = readCP;
	W[CP] = writeCP;
	R[C] = readC;
	W[C] = writeC;
	R[C_DEC] = c_dec;
	W[INC_C] = inc_c;
	R[C_INC] = c_inc;
	W[DEC_C] = dec_c;
	R[DP] = readDP;
	W[DP] = writeDP;
	R[D] = readD;
	W[D] = writeD;
	R[DEC_D] = dec_d;
	W[D_INC] = d_inc;
	R[INC_D] = inc_d;
	W[D_DEC] = d_dec;
}

void copy (address dst, address src) {
	instCount++;
	(*W[dst])((*R[src])());
}

void decode (word instruction) {
	copy((instruction>>12) & 0xf, (instruction>>8) & 0xf);
	copy((instruction>>4) & 0xf, instruction & 0xf);
}

void exec() {
	word instruction;
	
	instruction = RAM[Val[PC]];
	Val[PC]++;
	
	decode(instruction);
}
